### PR TITLE
feat(BREAKING): refactor cache handling to be more consistent

### LINKED ISSUE #19: [FEATURE] Cache 3xx redirects
# What / Why
It would be great if the same caching that is provided for content responses could be included for redirect responses as well using the same cache control header rules for the redirect response itself.

## When
Currently there is no cache or offline support for redirect requests as far as I can tell.

### LINKED ISSUE #25: [FEATURE] defaults() should do deep defaulting of headers
# What / Why

In `defaultedFetch` in `index.js` ([src](https://github.com/npm/make-fetch-happen/blob/latest/index.js#L44)), the defaulting is done with `Object.assign({}, _opts || {}, opts || {})`

This is a problem if `headers` are specified in the defaults *and* the invocation because the default `headers` object gets overridden.

## When

* n/a

## Where

* n/a

### Current Behavior

Take this example:

```
const customFetch = fetch.defaults({
    headers: {
        'X-Request-Id': uuid.v3(),
    }
});

customFetch(url, {
    headers: {
        Accept: 'application/json'
    }
}).then(...);
```
`Accept` is sent with the request but not `X-Request-Id`.

### Expected Behavior

Using the example above, I expect both `X-Request-Id` and `Accept` to be included in the headers sent with the request.

## Who

* n/a

## References

* n/a

### LINKED ISSUE #27: [QUESTION] Does ECONNRESET retry or not? Documentation says yes and no.
# What / Why

In the readme I see these two statements that seem contradictory:

> Retries will only be attempted on the following conditions ... Request errored with ECONNRESET ...

> The following are worth noting as explicitly not retried ... ECONNRESET currently has no support for restarting. ...

Is one of these a mistake? Or is "restarting" different than "retrying", and I'm not understanding?

From [this code](https://github.com/npm/make-fetch-happen/blob/c389a1e567327359aa4b34dbf805989ecbd9755b/index.js#L21-L28) it looks like `ECONNRESET` should be treated the same way as other connection errors. I'm hoping that's the case, and the second statement above is a mistake 😬

### LINKED ISSUE #28: [BUG] Response headers missing for 304 responses
# What / Why

make-fetch-happen deviates from browser caching behavior in the following ways:

1. When a 304 Not Modified response is received, and the cache is used for the response, the headers are missing from the [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response) object. Note that `.text()` and `.json()` still work on the response object and provide the cached response body -- it's just the headers that are missing.
2. A 304 Not Modified response status should be converted to a 200 before handed off to the user of make-fetch-happen. This is important for the [Response.ok](https://developer.mozilla.org/en-US/docs/Web/API/Response/ok) property to be true for responses that come from cache.

## When

From what I can tell, this occurs whenever a server responds with 304 and the cached version of the response is used. Generally, a server will send a 304 due to the `if-modified-since` or `if-none-match` headers.

## Where

https://github.com/systemjs/systemjs, which I help maintain, uses make-fetch-happen. I noticed this when requesting a resource from a webpack-dev-server, which properly responded with a 304 in accordance to the `if-modified-since` and `if-none-match` headers sent by make-fetch-happen.

### Current Behavior

Responses from cache are missing all response headers, indicate that they are not `ok`, and have the wrong HTTP status (304 instead of 200).

### Steps to Reproduce

See https://github.com/joeldenning/make-fetch-happen-304

### Expected Behavior

I expected make-fetch-happen to transparently convert the 304 into a 200 response that is `ok`, since that is how browsers implement fetch. If not that, I expected `res.headers.get('content-type')` to work.

## Who

Myself and @guybedford are interested in this one

## References

See https://github.com/systemjs/systemjs/pull/2171, where I have employed a workaround for this problem.

### LINKED ISSUE #30: [BUG] long npm install duration due to huge cacache entries
# What / Why

I have a cron job that runs every 5 minutes. It runs `npm install -g git3po`. And I've noticed that over time, this process is taking longer to complete. When re-running the same command manually, it takes 81 seconds to complete. For much of that time, I see "checking installable status" on screen. 

Running the same install with --verbose gives me the following output.... you can tell by the timestamps that each dependency takes ~1 second:

<details>

```
npm http fetch GET 304 https://registry.npmjs.org/lodash 7882ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/bluebird 8572ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/superagent 9271ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/chrono-node 9977ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/colors 12032ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/debug 12032ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/babel-runtime 12037ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/yargs 13436ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/js-yaml 13444ms (from cache)
```

</details>

I debugged this some, though not completely. It appears to reside in how make-fetch-happen leverages cacache. Here's what I saw in the cpu profile:

image

And zooming in:
Screen Shot 2020-06-19 at 2 19 48 PM

It's the [`cacache.get.info(this._path, key)`](https://github.com/npm/make-fetch-happen/blob/b04c4c16a7bb99ff2df1a3ee8ac5b738a43a1a30/cache.js#L42) call from m-f-s that results in the big latency. And why? Because I apparently have ~16,000 entries for some of my buckets. (That's 23MB of text) :) Doing some back-of-the-envelope math, if I have a new entry every 5 minutes, that'd be ~57 days ago, which seems roughly when I last noticed this and ended up doing `npm cache clean --force` (which did indeed resolve the problem).

Just for a sense of scale, my `~/.npm/cacache/index-v5` folder is now 2.8 GB. And I've confirmed it's only just the typical cacache text files in there.

So on the plus side, it doesn't appear I have a lot of duplicate files on disk due to smart caching. 🥳  But OTOH, the cache entries keeping track of all the kinds of requests that were satisfied by this cache entry keep growing.

Here's two entries out of the 16k, just for comparison:
image

### Full cache vs empty cache

To simulate empty cache I just renamed `~/.npm/cacache/index-v5`.

Full cache npm install -g duration: 81s
Empty cache npm install -g duration: 8s

### Steps to Reproduce

I haven't tried it, but I would imagine a script which just runs `git install -g whatever` in a loop would demonstrate the increased latency over time, especially as the entries exceed 10MB.

-----

👋 Isaac! Thanks for making this system and keeping it runnin! :D

### LINKED ISSUE #34: [BUG] http redirect makes integrity check failed
# What / Why

> http redirect like 302 will make integrity check failed

## Where

First it was found in `pacote`, but finally I locate the problem in `make-fetch-happen` package

### Steps to Reproduce

Use `pacote` with custom registry:

```js
const pacote = require( 'pacote' )

pacote.tarball( 'react@16.x', {
  registry: `https://registry.npm.taobao.org`
} ).then( buffer => console.log( 'got it', buffer ) )
```

## How

The root cause is that when requesting the tarball file, it will be 302 redirected to another location by my custom npm registry

e.g. visit https://registry.npm.taobao.org/react/download/react-16.13.1.tgz, it will be redirected to https://cdn.npm.taobao.org/react/-/react-16.13.1.tgz

the 302 request will use the unmatched integrity to do the integrity check, which will fail

### Current Behavior

Some error appears

```
sha1-xxx integrity checksum failed when using sha1: wanted sha1-yyy
```

### Expected Behavior

No Error

### LINKED ISSUE #39: Make `cacache` dependency optional
In the spirit of keeping the package lightweight, I'd like to suggest removing the `String` form argument of `cacheManager` and only accepting `Cache`. This would allow dropping the `cacache` dependency and reducing total module size by over 200kB which benefits users who don't need this feature.

### LINKED ISSUE #45: [FEATURE] Ignore the cache for normal POST requests
# What / Why

Ignore the cache for normal POST requests. because most POST request are inherently not cachable. To create less noise inside index-v5/1d/68

## When

When doing POST request , where the HTTP answer has not any cache specific headers and is not a redirect to a cacheable GET

## Where

```js
fetch(url, {
        method: `POST`,
        headers: {
            Accept: `text/html`,
            [`Content-Type`]: `application/json`,
            [`Accept-Language`]: options.lang,
        },
        body: JSON.stringify({}),
    })
```

### Current Behavior

The cache gets filled with 
`e6ec3ca5ada7fc4b516a043e4c0f1995c2e79889	{"key":"make-fetch-happen:request-cache:http://localhost:8000/x/u","integrity":"sha512-0wxO8Tl+lqMutURA76/b0THjA8j6o/1GFs2uwSt2iIWg26CxVzEr90MGVKO7ouE/WorbekoZC6Lblj37viA3Lw==","time":1610622806041,"size":1013,"metadata":{"url":"http://localhost:8000/x/u","reqHeaders":{"connection":["keep-alive"],"user-agent":["make-fetch-happen/8.0.10 (+https://npm.im/make-fetch-happen)"]},"resHeaders":{"content-type":["text/html"],"date":["Thu, 14 Jan 2021 11:13:26 GMT"],"connection":["keep-alive"],"keep-alive":["timeout=5"],"transfer-encoding":["chunked"],"x-fetch-attempts":["1"]}}}`

### Expected Behavior

Cache stays empty for POST request

### LINKED ISSUE #48: improve and extend caching semantics
today, make-fetch-happen handles some caching semantics reasonably well but has a flaw in that cached indexes can grow unbounded. we're also sending more requests (and downloading more data) than is necessary when we know the expected integrity of the content we're requesting.

we need to make the following modifications to improve the caching experience for npm

- [x] reduce the request and response headers stored in the index metadata to only those necessary for caching purposes
- [x] when an expected integrity is provided, disregard http caching semantics and ask cacache for the content directly. if cacache doesn't have the content, only then should we send the request
- [x] when sending a request to revalidate a cached response (typically due to max-age having expired), after receiving a 304 response and writing the new index, we must remove the original index matching our request

### PR DESCRIPTION
**Pull request by nlf**
**BREAKING CHANGE** the `cacheManager` option used to take either a
path, or an instance of an object with specific methods. it has been
deprecated in favor of `cachePath` which is strictly only a path,
and cacache is hard coded as the only caching store. If `cacheManager`
is provided, it will be copied to `cachePath` to ease the transition.

**BREAKING CHANGE** this module no longer deals with `warning`
headers, as they are deprecated anyway.

Refactored cache handling to more reliably store and lookup multiple
request/response pairs associated with the same URL. We also now
will skip writing data and indexes to the cache if the response fails its
ssri integrity stream, as well as deduplicate and prune index entries
with each read, which will help ensure that our cache does not grow
totally unbounded.

This change ended up a lot bigger than I had hoped due to how things were organized previously. I think this refactor breaks things up in a reasonable way so that it avoids circular dependencies while not wedging all of the logic in one file.

I kept as many of the existing tests as made sense, though many were refactored. I finished up a couple of TODOs that were in the tests too.

## References

Closes npm/statusboard#357
Depends on npm/cacache#55
Closes #48 
Closes #45 
Closes #39 (by doing the opposite of making cacache required)
Closes #34 
Fixes #30 
Fixes #28 
Closes #27 
Closes #26
Fixes #25
Closes #23 
Fixes #19

### CODE REVIEW

**Comment by nlf** [comment]
build is failing because for some reason CI isn't able to clone the cacache repo, that'll be fixed when npm/cacache#53 lands and this is updated to use the published dep

**Comment by nlf** [comment]
added invalidation after a non-GET/HEAD request to the same url, and made sure that null integrity values are handled correctly (i.e. the old null values that signify a delete are ignored and removed from the index) so this pull request now depends on npm/cacache#55 landing

File: lib/cache/policy.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  }
+
+  // returns true if the request and response allow caching
+  storable () {
+    return this.policy.storable()
+  }
+
+  // NOTE: this is a hack to avoid parsing the cache-control
+  // header ourselves, it returns true if the response's
+  // cache-control contiains must-revalidate
```
Comment (wraithgar) [inline lib/cache/policy.js:112]:
```suggestion
  // cache-control contains must-revalidate
```

**Review by wraithgar — APPROVED** [review]
Walked through this w/ @nlf, code layout makes sense and I'm excited for the performance increases we get by caching better.

**Review by wraithgar — APPROVED** [review]
Race condition making the tests flaky was fixed. Yay more bugs fixed.

**Pull request merged**